protocol/x/accountplus/ante/ante.go

```diff
@@ -0,0 +1,299 @@
+					"failed to get initialized authenticator "+
+						"(account = %s, authenticator id = %d, msg index = %d, msg type url = %s)",
+					account,
+					selectedAuthenticatorId,
+					msgIndex,
+					sdk.MsgTypeURL(msg),
+				)
+		}
+
+		// Generate the authentication request data
+		authenticationRequest, err := authenticator.GenerateAuthenticationRequest(
+			ctx,
+			ad.cdc,
+			ad.accountKeeper,
+			ad.sigModeHandler,
+			account,
+			feePayer,
+			feeGranter,
+			fee,
+			msg,
+			tx,
+			msgIndex,
+			simulate,
+		)
+		if err != nil {
+			return sdk.Context{},
+				errorsmod.Wrapf(
+					err,
+					"failed to generate authentication data "+
+						"(account = %s, authenticator id = %d, msg index = %d, msg type url = %s)",
+					account,
+					selectedAuthenticator.Id,
+					msgIndex,
+					sdk.MsgTypeURL(msg),
+				)
+		}
+
+		a11r := selectedAuthenticator.Authenticator
+		stringId := strconv.FormatUint(selectedAuthenticator.Id, 10)
+		authenticationRequest.AuthenticatorId = stringId
+
+		// Consume the authenticator's static gas
+		ctx.GasMeter().ConsumeGas(a11r.StaticGas(), "authenticator static gas")
+
+		// Authenticate should never modify state. That's what track is for
+		neverWriteCtx, _ := ctx.CacheContext()
+		authErr := a11r.Authenticate(neverWriteCtx, authenticationRequest)
+
+		// If authentication is successful, continue
+		if authErr == nil {
+			// Append the track closure to be called after every message is authenticated
+			// Note: pre-initialize type URL to avoid closure issues from passing a msg
+			// loop variable inside the closure.
+			currentMsgTypeURL := sdk.MsgTypeURL(msg)
+			tracks = append(tracks, func() error {
+				err := a11r.Track(ctx, authenticationRequest)
+				if err != nil {
+					// track should not fail in normal circumstances,
+					// since it is intended to update track state before execution.
+					// If it does fail, we log the error.
+					metrics.IncrCounter(metrics.AuthenticatorTrackFailed, 1)
+					ad.accountPlusKeeper.Logger(ctx).Error(
+						"track failed",
+						"account", account,
+						"feePayer", feePayer,
+						"msg", currentMsgTypeURL,
+						"authenticatorId", stringId,
+						"error", err,
+					)
+
+					return errorsmod.Wrapf(
+						err,
+						"track failed (account = %s, authenticator id = %s, authenticator type, %s, msg index = %d)",
+						account,
+						stringId,
+						a11r.Type(),
+						msgIndex,
+					)
+				}
+				return nil
+			})
+		}
+
+		// If authentication failed, return an error
+		if authErr != nil {
+			return ctx, errorsmod.Wrapf(
+				authErr,
+				"authentication failed for message %d, authenticator id %d, type %s",
+				msgIndex,
+				selectedAuthenticator.Id,
+				selectedAuthenticator.Authenticator.Type(),
+			)
+		}
+	}
+
+	// If the transaction has been authenticated, we call Track(...) on every message
+	// to notify its authenticator so that it can handle any state updates.
+	for _, track := range tracks {
+		if err := track(); err != nil {
```

Comment:
Question: I see our current authenticators either call `subTrack` recursively or is empty. What's an actual use case of this?

Reply:
We don't need it right now for our current use case since our authentications are all stateless / static (public key signature verification, message filter, clob pair id, subaccount number).

This persists any state changes that is required by the authenticator. e.g. if we want to support something like `You can place up to X orders`, we would be using the `track` method to increment a counter in state

Reply:
also down to get rid of this for v7.1 for simplicity

Reply:
I see, no objections to keeping this in -  IIUC it doesn't increase the attack vector since no implemented authenticators implement tracking